### PR TITLE
Allow custom key separators for JSON based file formats and exporters.

### PR DESCRIPTION
**Pull request by mkarol-neurosys**
## Proposed changes
During discussion in https://github.com/translate/translate/pull/5179 I realized that Weblate during export is losing key structure by not escaping dots in the natural language keys. Instead of escaping and unescaping separators throughout whole application I've implemented support for custom key separator allowing for simple import and export of files containing non-default key separators.

## Checklist
- [x] Lint and unit tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added documentation to describe my feature.
- [x] I have squashed my commits into logic units.
- [x] I have described the changes in the commit messages.

## Other information
Setting up key separator when creating component from scratch. Option is also available when creating component from other sources as well it is possible to change key separator in component settings.
image

Have a great day!
Michał Karol

### CODE REVIEW

File: weblate/templates/component.html [hunk 1]
```diff
@@ -98,7 +98,6 @@
       {% if user_can_edit_component %}
       <li><a href="{% url 'guide' path=object.get_url_path %}">{% trans "Community localization checklist" %}</a></li>
       <li><a href="{% url 'addons' path=object.get_url_path %}">{% trans "Add-ons" %}</a></li>
-      <li><a href="{% url 'settings' path=object.get_url_path %}">{% trans "Settings" %}</a></li>
```
Comment (nijel) [inline weblate/templates/component.html:101]: This seems wrong.

**Review by nijel — COMMENTED** [review]
Thanks for your contribution, unfortunately I see several issues with it:

* The setting is done at the component level, but doesn't affect file loading when parsing files from the repository. That is confusing.
* When introducing parameters for file formats, we would rather need something more generic, not adding arbitrary parameters for few formats. See https://github.com/WeblateOrg/weblate/issues/1897
* Hard-coding separator to `.` would break YAML as it historically uses `->`, see https://github.com/translate/translate/blob/fea03126d2421454bb6d7f029eff2342c681dc30/translate/storage/yaml.py#L32 (I see you've limited this to JSON subclasses only, but this makes the setting behavior inconsistent)
* Generally, Weblate preserves whatever is in the file in the round-trip. What is the motivation to add this specific exporter and not just save changed files within Weblate?

File: weblate/templates/component.html (hunk 1 above)
Comment (mkarol-neurosys) [reply weblate/templates/component.html:101]: True. It should be other PR. In component overview you can have Settings menu item just right next to other Setttings menu item (from line 106), both pointing to the same url.

Comment (nijel) [reply weblate/templates/component.html:101]: Ah, I see, fixed that in 76379e559e051536d5617209f9009f20ef6fbd38.

**Comment by mkarol-neurosys** [comment]
Ad 1) When loading from vsc, after setting the source, there should be and option to specify key separator on `ComponentCreateForm`. Later, when celery task is performing `weblate/trans/models/component.py:2448 check_sync`  function it uses `weblate/trans/models/translation.py:268 self.component.file_format_cls` which uses wrapped file format class from component.

Ad 2) This is currently rather quick fix, I am planning to add changes to `translate` to allow for key separator customization.  If you want I can start with changes to the `translate` and utilizing those changes in weblate later on.

Ad 3) Like I said, it is first step. 

Ad 4) Unfortunately when some edits are made to the translations, there is no possibility to export changed i18nextv4 translations. JSON and JSON nested formats are stripping out pluras, and i18nextv4 exporter alone will nest keys in resulting JSON regardless whether the dot was inside the key eg. `{ "Natural language key. Will be nested": "some translation" }`.

**Comment by nijel** [comment]
> Unfortunately when some edits are made to the translations, there is no possibility to export changed i18nextv4 translations.

You don't need to convert the format via exporters when changing existing i18nextv4 translation. Just download the file as it is. The format exporting is there for interoperability with a few formats, and not as a generic framework for converting anything to anything.

**Comment by mkarol-neurosys** [comment]
Unfortunately there is no other way to download only reviewed and accepted translations. Exporters allow for selecting subset of translations without downloading whole file.

**Comment by nijel** [comment]
I understand your need, but I see this more like a workaround than a solution.

* We need a generic solution for file format parameters, adding a database field for a single parameter for a few file formats is not going to scale. This is what https://github.com/WeblateOrg/weblate/issues/1897 describes.
* There is https://github.com/WeblateOrg/weblate/issues/3745 to let Weblate commit only approved strings what might be a better approach to your problem.

**Pull request closed without merging**